ee/hogai/tools/read_data.py

```diff
@@ -53,35 +62,76 @@
 Suggest the user to contact the admins.
 """.strip()
 
-ReadDataKind = Literal["datawarehouse_schema"]
-ReadDataAdminAccessKind = Literal["datawarehouse_schema", "billing_info"]
-ReadDataWithArtifactsKind = Literal["datawarehouse_schema", "artifacts"]
-ReadDataAdminAccessWithArtifactsKind = Literal["datawarehouse_schema", "billing_info", "artifacts"]
+INSIGHT_NOT_FOUND_PROMPT = """
+The insight with the ID "{short_id}" was not found or uses an unsupported query type. Please verify the insight ID is correct.
+""".strip()
+
+INSIGHT_RESULT_TEMPLATE = """
+Name: {{{insight_name}}}
+Insight ID: {{{insight_id}}}
+{{#description}}
+Description: {{{description}}}
+{{/description}}
+
+{{{results}}}
+""".strip()
+
+INSIGHT_SCHEMA_TEMPLATE = """
+Name: {{{insight_name}}}
+Insight ID: {{{insight_id}}}
+{{#description}}
+Description: {{{description}}}
+{{/description}}
+
+Query definition:
+```json
+{{{query_schema}}}
+```
+""".strip()
+
+
+class ReadDataWarehouseSchema(BaseModel):
+    """Returns the SQL ClickHouse schema for the user's data warehouse."""
+
+    kind: Literal["datawarehouse_schema"] = "datawarehouse_schema"
+
 
-KindUnion = ReadDataWithArtifactsKind | ReadDataAdminAccessWithArtifactsKind | ReadDataKind | ReadDataAdminAccessKind
+class ReadInsight(BaseModel):
+    """Retrieves an existing saved insight by its short ID."""
 
+    kind: Literal["insight"] = "insight"
+    insight_id: str = Field(description="The short ID of the insight (found in URLs like /insights/abc123).")
+    execute: bool = Field(
+        default=False,
+        description="If true, executes the insight query and returns results. If false, returns only the insight definition.",
```

Comment:
Why are we doing if/else here, I think the query should be shown on execution too, something like: "I ran this query, I got these results". If we don't do this the agent might execute the insight, doubt the results, run this tool again to check the query, etc. doing extra loops for nothing. You would basically always get the definition, and optionally execute it to get the results. Unless you have some clear indication why it's better to separate the two.

Reply:
Ok fair